### ISSUE TITLE
Color serialization with channel keywords and calc

### ISSUE THREAD

**Issue by tiaanl** [web-platform-tests/wpt#47921]
When specifying a channel keyword in a color there are two methods:

`rgb(from red r g b)` and `rgb(from red calc(r) calc(g) calc(b))`

According to [css/css-color-parsing/color-valid-relative-color.html](https://github.com/web-platform-tests/wpt/blob/41a06b4bedff395c2d80170892a57471d9eb05a3/css/css-color/parsing/color-valid-relative-color.html) they serialize exactly as specified, so specifying a channel keyword _without_ calc serializes _without_ calc and specifying _with_ calc serializes _with_ calc.

When a calc contains a single numerical value, the `calc()` part can be omitted, otherwise it should be serialized with it as per [https://drafts.csswg.org/css-values-4/#calc-serialize](https://drafts.csswg.org/css-values-4/#calc-serialize).

Should these rules apply to channel keywords as well and how would it change the examples I gave?

**Comment by nt1m** [comment]
@weinig

**Comment by tiaanl** [comment]
@svgeesus

**Comment by svgeesus** [comment]
From CSS Color 4, [14.1. Resolving sRGB values](https://drafts.csswg.org/css-color-4/#resolving-sRGB-values):

> For historical reasons, when [calc()](https://drafts.csswg.org/css-values-4/#funcdef-calc) in sRGB colors resolves to a single value, the declared value serialises without the "calc(" ")" wrapper.

and from CSS Color 5, [11.2. Serializing Relative Color Functions](https://drafts.csswg.org/css-color-5/#serial-relative-color):

> The serialization of the declared value of a relative color function is a string identifying the color function in all-lowercase, followed by "(from ", followed by the serialization of the declared value of the origin color, followed by a single space, followed by a singly-space-separated list of the arguments to the color function, followed by ")".

I take that to mean that, for a relative sRGB color, any `calc` in the origin color goes away and is replaced by the numerical result. I agree that no specific mention is made for `calc` in the arguments but I would assume that the same rule applies. So I would expect the channel keywords to not contain `calc`.

@weining your tests clearly expect `calc` to be preserved, although perhaps the spec said something different at the time you wrote them?

**Comment by tiaanl** [comment]
Would this also apply to other color syntax as the spec specifically mention `sRGB`?

`hsl(from rebeccapurple calc(h) calc(s) calc(l))` == `hsl(from rebeccapurple h s l)`

or

`oklch(from oklch(0.7 0.45 30) calc(l) calc(c) calc(h))` == `oklch(from oklch(0.7 0.45 30) l c h)`

**Comment by svgeesus** [comment]
Also [from the spec](https://drafts.csswg.org/css-color-4/#resolving-sRGB-values):

> This applies to:

>    [hex colors](https://drafts.csswg.org/css-color-4/#hex-color)
>    [rgb()](https://drafts.csswg.org/css-color-4/#funcdef-rgb) and [rgba()](https://drafts.csswg.org/css-color-4/#funcdef-rgba) values
>    [hsl()](https://drafts.csswg.org/css-color-4/#funcdef-hsl) and [hsla()](https://drafts.csswg.org/css-color-4/#funcdef-hsla) values
>    [hwb()](https://drafts.csswg.org/css-color-4/#funcdef-hwb) values
>    [named colors](https://drafts.csswg.org/css-color-4/#named-color)
>    [system colors](https://drafts.csswg.org/css-color-4/#css-system-colors)
>    [deprecated-colors](https://drafts.csswg.org/css-color-4/#deprecated-system-colors)

> It does not apply to:

>    [color()](https://drafts.csswg.org/css-color-4/#funcdef-color) values using the [srgb](https://drafts.csswg.org/css-color-4/#valdef-color-srgb) or [srgb-linear](https://drafts.csswg.org/css-color-4/#valdef-color-srgb-linear) [color space](https://drafts.csswg.org/css-color-4/#color-space)s.

So yes to the first one (`hsl` is an sRGB color) and no to the second one.

IIRC the "serialises without the `"calc(" ")"` wrapper." is historical, for backwards compat, and so does nort apply the the non-sRGB syntaxes.

**Comment by yisibl** [comment]
@svgeesus The conclusion here appears to be inconsistent with the current WPT and specifications.

https://github.com/web-platform-tests/wpt/blob/144460ebf104adf40ef7b991919107e02f0839f1/css/css-color/parsing/color-valid-relative-color.html#L100

The `calc()` function is still included in the expected value here.

Furthermore, the examples in the CSS specification also retain `calc()`.

> For example, the serialization of the declared value of
rgb(from red calc(r / 2) g calc(30%));
is the string "rgb(from red calc(0.5 * r) g calc(30%))", while the serialization of the computed value is the string "color(srgb 0.5 0 0.3)".
> https://drafts.csswg.org/css-color-5/#ex-serial-rcs-calc
> https://github.com/w3c/csswg-drafts/commit/9cc2a6b39de8de9211e0fe524bb1c82e533955a4

**Comment by tiaanl** [comment]
Yes, right now we're serializing all `calc()` with only a color component inside with the `calc()` wrapper, including `rgb`, `hsl` and `hwb` (srgb formats).

`rgb(from red calc(r) calc(g) calc(b))` -> `rgb(from red calc(r) calc(g) calc(b))`

but should really be `rgb(from red r g b)`.

I don't think this is related to `calc()` with uncomputed values inside, like `calc(r / 2)`.  That should serialize the calc in specified values regardless.

So we need to update tests for `rgb`, `hsl` and `hwb` with relative syntax.

**Comment by svgeesus** [comment]
OK, any volunteers to update the relevant WPT? Happy to review, ping here when ready.
